File: C++/extend.cpp
#include <iostream>
#include "extend.hpp"
#include "smith_waterman.hpp"

#define MIN(v1, v2) (((v1) > (v2)) ? (v2) : (v1))
#define MAX(v1, v2) (((v1) < (v2)) ? (v2) : (v1))

std::string _extend_and_score(AdjacentPair pair,
                             std::string query,
                             std::string data,
                             int match,
                             int mismatch,
                             int gap,
                             int minscore,
                             bool score,
                             bool printing)
{
    // find left most indices
    int dleftindex  = MIN(pair.dindex1, pair.dindex2);
    int qleftindex  = MIN(pair.qindex1, pair.qindex2);
    int drightindex = MAX(pair.dindex1, pair.dindex2);
    int qrightindex = MAX(pair.qindex1, pair.qindex2);

    // build string
    std::string qextended = query.substr(qleftindex, pair.length);
    std::string dextended = data.substr(dleftindex, pair.length);

    // extend left
    int qexindex = qleftindex;
    int dexindex = dleftindex;
    while (qexindex - 1 >= 0 && dexindex - 1 >= 0) {
		qexindex--; dexindex--;
        qextended = query[qexindex] + qextended;
        dextended = data[dexindex] + dextended;
        if (score
            && minscore > smith_waterman(qextended, dextended, match, mismatch, gap, true))
        {
            return "NULL";
        }
    }
	
    // extend left pair to the right
    qexindex = qleftindex + pair.length - 1;
    dexindex = dleftindex + pair.length - 1;
    while (qexindex + 1 < qrightindex) {
		qexindex++; dexindex++;
        qextended = qextended + query[qexindex];
        dextended = dextended + data[dexindex];
    }

    // extend right with gaps until qextended aligns with data
    while (dexindex + 1 < drightindex) {
		qexindex++; dexindex++;
        qextended = qextended + "-";
        dextended = dextended + data[dexindex];
    }

    // append the right pair
    qextended = qextended + query.substr(qrightindex, pair.length);
    dextended = dextended + data.substr(drightindex, pair.length);

    // extend right
    qexindex = qrightindex + pair.length - 1;
    dexindex = drightindex + pair.length - 1;
    while (qexindex + 1 < query.size() && dexindex + 1 < data.size()) {
		qexindex++; dexindex++;
        qextended = qextended + query[qexindex];
        dextended = dextended + data[dexindex];
        if (score
            && minscore < smith_waterman(qextended, dextended, match, mismatch, gap, true))
        {
            return "NULL";
        }
    }

    if (printing) {
        std::cout << "Data Ext:\t" << dextended << std::endl;
        std::cout << "Quer Ext:\t" << qextended << std::endl;
    }
    return qextended;
}

void test_extend()
{
    std::string query = "GTCTGAACTGAGC";
    std::string data  = "AGTCTGATGACTGGGGAACTCGA";
    std::string word1 = "TC";
    std::string word2 = "CT";

    int qindex1 = query.find(word1);
    int dindex1 = data.find(word1);
    int qindex2 = query.find(word2, qindex1 + word1.size());
    int dindex2 = data.find(word2, dindex1 + word1.size());

    AdjacentPair pair(word1, word2, dindex1, qindex1, dindex2, qindex2);

    std::cout << "Query:\t\t" << query << std::endl;
    std::cout << "Data:\t\t" << data << std::endl;

    int match = 2;
    int mismatch = -1;
    int gap = -1;
    int minscore = 6;
    bool score = true;
    bool printing = true;
    std::string result = _extend_and_score(pair,
                                           query,
                                           data,
                                           match, mismatch, gap,
                                           minscore,
                                           score,
                                           printing);
}

File: python/lib/extend.py
from copy import copy
from typing import Dict, List

from smith_waterman import _smith_waterman
from pairs import AdjacentPair

"""
Internal
"""

class Extended:
    def __init__(self, dindex, extended_pair):
        """
        @brief: Hold the extended pair and where it was found in the database.
        """
        self.dindex = dindex
        self.extended_pair = extended_pair

"""
2 adjacent pairs
LEFT (lesser index)
    extend leftward one at a time and score
MIDDLE
    fill with gap characters
RIGHT (greater index)
    extend rightward one at a time and score
"""

def extend_and_score(pair: AdjacentPair,
                     query: str,
                     data: str,
                     match: int,
                     mismatch: int,
                     gap: int,
                     minscore: int,
                     score=False,
                     printing=True) -> str:
    """
    @brief: Extend an adjacent pair while scoring each extension. Stop extending if the extended word \\
            scores too low. \\
    @param pair:     The adjacent pair of words and their indices. \\
    @param query:    The section the adjacent pair is from. \\
    @param data:     The section the adjacent pair is going to be extended to. \\
    @param match:    Smith Waterman score for a match. \\
    @param mismatch: Smith Waterman score for a mismatch. \\
    @param gap:      Smith Waterman score for a gap. \\
    @param minscore: Minimum Smith Waterman score before the extending stops. \\
    @param score:    Perform Smith Waterman scoring on true. \\
    @param printing: Print the data after extending on true. \\
    @return: The string containing both pairs extended to the query from the database. \\
    """
    # find left-most indices
    dleftindex = min([pair.dindex1, pair.dindex2])
    qleftindex = min([pair.qindex1, pair.qindex2])
    drightindex = max([pair.dindex1, pair.dindex2])
    qrightindex = max([pair.qindex1, pair.qindex2])

    # build string
    qextended = query[qleftindex:qleftindex + pair.length]
    dextended = data[dleftindex:dleftindex + pair.length]

    # extend left
    qexindex = copy(qleftindex)
    dexindex = copy(dleftindex)
    while qexindex - 1 >= 0 and dexindex - 1 >= 0:
        qexindex -= 1
        dexindex -= 1
        qextended = query[qexindex] + qextended
        dextended = data[dexindex] + dextended
        if score:
            s = _smith_waterman(qextended, dextended, match=match, mismatch=mismatch, gap=gap, just_score=True)
            if s < minscore:
                return None
    
    # extend left pair to the right
    qexindex = qleftindex + pair.length - 1
    dexindex = dleftindex + pair.length - 1
    while qexindex + 1 < qrightindex:
        qexindex += 1
        dexindex += 1
        qextended = qextended + query[qexindex]
        dextended = dextended + data[dexindex]
    
    # extend right with gaps until qextended aligns with data
    while dexindex + 1 < drightindex:
        qexindex += 1
        dexindex += 1
        qextended = qextended + '-'
        dextended = dextended + data[dexindex]
    
    # append the right pair
    qextended = qextended + query[qrightindex:qrightindex + pair.length]
    dextended = dextended + data[drightindex:drightindex + pair.length]
    
    # extend right
    qexindex = qrightindex + pair.length - 1
    dexindex = drightindex + pair.length - 1
    while qexindex + 1 < len(query) and dexindex + 1 < len(data):
        qexindex += 1
        dexindex += 1
        qextended = qextended + query[qexindex]
        dextended = dextended + data[dexindex]
        if score:
            s = _smith_waterman(qextended, dextended, match=match, mismatch=mismatch, gap=gap, just_score=True)
            if s < minscore:
                return None
    
    if printing:
        print(f"Data Ext:\t{dextended}")
        print(f"Quer Ext:\t{qextended}")
    
    return qextended

"""
External
"""

def extend_filter(pairs: Dict[str, Dict[str, List[AdjacentPair]]],
                  minscore: int,
                  match: int,
                  mismatch: int,
                  gap: int) -> Dict[str, Dict[str, List[Extended]]]:
    pass
